plugins/inputs/pubsub/pubsub.go

```diff
@@ -0,0 +1,290 @@
+// canceling the context and waits for goroutines to finish.
+func (ps *PubSub) Stop() {
+	ps.cancel()
+	ps.wg.Wait()
+}
+
+// Gather does nothing for this service input.
+func (ps *PubSub) Gather(acc telegraf.Accumulator) error {
+	return nil
+}
+
+// SetParser implements ParserInput interface.
+func (ps *PubSub) SetParser(parser parsers.Parser) {
+	ps.parser = parser
+}
+
+// onMessage handles parsing and adding a received message to the accumulator.
+func (ps *PubSub) onMessage(ctx context.Context, msg *pubsub.Message) error {
+	if ps.MaxMessageLen > 0 && len(msg.Data) > ps.MaxMessageLen {
+		return fmt.Errorf("message longer than max_message_len (%d > %d)", len(msg.Data), ps.MaxMessageLen)
+	}
+	metrics, err := ps.parser.Parse(msg.Data)
+	if err != nil {
+		msg.Ack()
+		return err
+	}
+	if len(metrics) == 0 {
+		msg.Ack()
+		return nil
+	}
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case ps.sem <- empty{}:
+		break
+	}
+
+	ps.mu.Lock()
+	defer ps.mu.Unlock()
+	id := ps.acc.AddTrackingMetricGroup(metrics)
+	ps.undeliveredMessages[id] = msg
+
+	return nil
+}
+
+func (ps *PubSub) receiveDelivered(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case info := <-ps.acc.Delivered():
+			msg := ps.handleDelivered(info.ID())
+			if msg != nil {
+				msg.Ack()
+			}
+		}
+	}
+}
+
+func (ps *PubSub) handleDelivered(id telegraf.TrackingID) *pubsub.Message {
+	ps.mu.Lock()
+	defer ps.mu.Unlock()
+
+	msg, ok := ps.undeliveredMessages[id]
+	if !ok {
+		return nil
+	}
+	delete(ps.undeliveredMessages, id)
+	<-ps.sem
+	return msg
+}
+
+func (ps *PubSub) initPubSubHandles(ctx context.Context) error {
+	var credsOpt option.ClientOption
+	if ps.CredentialsFile != "" {
+		credsOpt = option.WithCredentialsFile(ps.CredentialsFile)
+	} else {
+		creds, err := google.FindDefaultCredentials(ctx, pubsub.ScopeCloudPlatform)
+		if err != nil {
+			return fmt.Errorf("credentials JSON file not provided and unable to find GCP Application Default Credentials: %v", err)
+		}
+		credsOpt = option.WithCredentials(creds)
+	}
+
+	client, err := pubsub.NewClient(
+		ctx,
+		ps.Project,
+		credsOpt,
+		option.WithScopes(pubsub.ScopeCloudPlatform),
+		option.WithUserAgent(gcpUserAgentString()))
+	if err != nil {
+		return fmt.Errorf("unable to generate PubSub client: %v", err)
+	}
+
+	ps.client = client
+	ps.subRefs = make(map[string]*pubsub.Subscription)
+	for _, subId := range ps.Subscriptions {
+		subRef := ps.client.Subscription(subId)
+		subRef.ReceiveSettings = pubsub.ReceiveSettings{
+			NumGoroutines:          ps.MaxReceiverGoRoutines,
+			MaxExtension:           time.Duration(ps.MaxExtensionSec) * time.Second,
+			MaxOutstandingMessages: ps.MaxOutstandingMessages,
+			MaxOutstandingBytes:    ps.MaxOutstandingBytes,
+		}
+		ps.subRefs[subRef.ID()] = subRef
+	}
+
+	return nil
+}
+
+func gcpUserAgentString() string {
+	return fmt.Sprintf("InfluxData Telegraf/%s (%s)", internal.Version(), runtime.Version())
```

Comment:
I was hoping to reuse this as well as some of the client setup code for the output plugin. It might make sense to also use this in existing GCP plugins (i.e. Stackdriver). Is there somewhere I can add shared utils?

Reply:
We have [something similar](https://github.com/influxdata/telegraf/blob/master/plugins/outputs/influxdb/http.go#L139) in a few plugins such as the influxdb output.  If you like you can add a function to `internal/internal.go` `func ProductToken() string` and have it return `Telegraf/1.9.1`.  

The idea here is to send a [http product token](https://tools.ietf.org/html/rfc2616#section-3.8), I wouldn't mind also including the Go version: `Telegraf/1.9.1 Go/1.11.3` however I don't see a very good way to get the Go version string into this format.